### PR TITLE
Refactor swap module

### PR DESCRIPTION
**Pull request by herryho**
swap moudle

### CODE REVIEW

File: brml/swap/src/lib.rs [hunk 1]
```diff
@@ -1,1022 +1,1400 @@
...
+				// get the total pool token shares for the specific pool
+				let all_pool_tokens = PoolTokensInPool::<T>::get(pool_id);
+				// Asset id
+				let token_id = tk.0;
+				// get the user's balance for a specific token
+				let user_token_pool_balance = T::AssetTrait::get_account_asset(token_id, &provider).balance;
+				// the balance of a specific token in a pool
+				let token_pool_balance = TokenBalancesInPool::<T>::get(pool_id, token_id);
+				// the amount of the token that the user should deposit
+				let should_deposit_amount = token_pool_balance * new_pool_token / all_pool_tokens;
```
Comment (Dengjianping) [inline brml/swap/src/lib.rs:286]: * shoule be replaced by saturating_mul

File: brml/swap/src/lib.rs [hunk 2]
```diff
@@ -1,1022 +1,1400 @@
...
+		// ****************************************************************************
+		/// A user adds liquidity by depositing only one kind of token.
+		/// So we need to calculate the corresponding pool token share the user should get.
+		/// (add liquidity)(single asset) given share in => amount out
+		#[weight = T::DbWeight::get().reads_writes(1, 1)]
+		fn add_single_liquidity_given_shares_in(
+			origin,
+			pool_id: T::PoolId,
+			asset_id: TokenSymbol,
+			new_pool_token: T::Balance,
```
Comment (Dengjianping) [inline brml/swap/src/lib.rs:372]: Add this attribute: #[compact]

File: brml/swap/src/lib.rs [hunk 3]
```diff
@@ -1,1022 +1,1400 @@
...
+
+		// ****************************************************************************
+		/// User remove liquidity with only one kind of token
+		/// (remove liquidity)(single asset) given amount in => shares out
+		#[weight = T::DbWeight::get().reads_writes(1, 1)]
+		fn remove_single_asset_liquidity_given_amount_in(
+			origin,
+			pool_id: T::PoolId,
+			asset_id: TokenSymbol,
+			token_amount: T::Balance  // The number of out-token that the user want to remove liquidity with from the pool.
```
Comment (Dengjianping) [inline brml/swap/src/lib.rs:476]: Add this attribute: #[compact]

File: brml/swap/src/lib.rs [hunk 4]
```diff
@@ -1,1022 +1,1400 @@
...
+			// Meanwhile, issue money to user's account and deducted from the pool.
+			for tk in token_balances_in_pool_iter {  //0 position is asset_id, 1 position is balance
+				// get the total pool token shares for the specific pool
+				let all_pool_tokens = PoolTokensInPool::<T>::get(pool_id);
+				// Asset id
+				let token_id = tk.0;
+				// the balance of a specific token in a pool
+				let token_pool_balance = TokenBalancesInPool::<T>::get(pool_id, token_id);
+				// the amount of the token that the user should deposit
+				let can_withdraw_amount = token_pool_balance * pool_amount_out / all_pool_tokens;
```
Comment (Dengjianping) [inline brml/swap/src/lib.rs:553]: '*' shoule be replaced by saturating_mul

File: brml/swap/src/lib.rs [hunk 5]
```diff
@@ -1,1022 +1,1400 @@
...
+			// ensure all the elements of the tokenForPoolMap are ok.
+			for token_info in map_iter_1 {
+				ensure!(T::AssetTrait::token_exists(token_info.token_id), Error::<T>::TokenNotExist);
+				ensure!(token_info.token_balance > Zero::zero(), Error::<T>::AmountBelowZero);
+
+				// get the user's balance for a specific token
+				let user_token_balance = T::AssetTrait::get_account_asset(token_info.token_id, &creator).balance;
+				ensure!(user_token_balance >= token_info.token_balance, Error::<T>::NotEnoughBalance);
+				// Add up the total weight
+				total_weight += token_info.token_weight;
```
Comment (Dengjianping) [inline brml/swap/src/lib.rs:769]: '+' shoule be replaced by saturating_add

File: brml/swap/src/lib.rs [hunk 6]
```diff
@@ -1,1022 +1,1400 @@
...
+			Pools::<T>::insert(new_pool_id, new_pool);
+
+			let map_iter_2 = token_for_pool_vec.iter();
+			// initialize the pool
+			for token_info in map_iter_2 {
+				// destroy user's token
+				T::AssetTrait::asset_redeem(token_info.token_id, &creator, token_info.token_balance);
+
+				// insert TokenWeightsInPool
+				let token_normalized_weight = token_info.token_weight * T::WeightPrecision::get() / total_weight;
```
Comment (Dengjianping) [inline brml/swap/src/lib.rs:791]: '*' shoule be replaced by saturating_mul

File: brml/swap/src/lib.rs [hunk 7]
```diff
@@ -1,1022 +1,1400 @@
...
+
+			Ok(())
+		}
+
+		/// set the pool status to be true or false.
+		#[weight = T::DbWeight::get().reads_writes(1, 1)]
+		pub fn set_pool_status(
+			origin,
+			pool_id: T::PoolId,
+			new_status: bool) -> DispatchResult {
```
Comment (Dengjianping) [inline brml/swap/src/lib.rs:825]: Formater issue
```rust
new_status: bool
) -> DispatchResult {
```

File: brml/swap/src/lib.rs [hunk 8]
```diff
@@ -1,1022 +1,1400 @@
...
+
+	// ********************************************************************************
+	// below are the interfaces needed from other pallets.
+	// Query for the current bonus balance for the pool
+	pub(crate) fn get_bonus_pool_balance(_pool_id: T::PoolId) -> T::Balance {
+		T::Balance::from(1_000_000) // to get from other pallets. Not yet implemented
+	}
+}
+
+#[allow(dead_code)]
```
Comment (Dengjianping) [inline brml/swap/src/lib.rs:1385]: Removing line 1385 to 1400 would be better.
It's already deprecated.

**Review by Dengjianping — COMMENTED** [review]
Some comments have been added.

File: brml/swap/src/lib.rs (hunk 1 above)
Comment (herryho) [reply brml/swap/src/lib.rs]: fixed.

File: brml/swap/src/lib.rs (hunk 3 above)
Comment (herryho) [reply brml/swap/src/lib.rs]: fixed.

File: brml/swap/src/lib.rs (hunk 5 above)
Comment (herryho) [reply brml/swap/src/lib.rs]: fixed.

File: brml/swap/src/lib.rs (hunk 4 above)
Comment (herryho) [reply brml/swap/src/lib.rs]: fixed.

File: brml/swap/src/lib.rs (hunk 2 above)
Comment (herryho) [reply brml/swap/src/lib.rs]: fixed.

File: brml/swap/src/lib.rs (hunk 6 above)
Comment (herryho) [reply brml/swap/src/lib.rs]: fixed.

File: brml/swap/src/lib.rs (hunk 7 above)
Comment (herryho) [reply brml/swap/src/lib.rs]: fixed.

File: brml/swap/src/lib.rs (hunk 8 above)
Comment (herryho) [reply brml/swap/src/lib.rs]: fixed.

**Pull request merged**